services/brig/src/Brig/API.hs

```diff
@@ -1270,33 +1270,50 @@ listUsers self = \case
     byIds uids = lift $ API.lookupProfiles self uids
 
 listActivatedAccountsH :: JSON ::: Either (List UserId) (List Handle) -> Handler Response
-listActivatedAccountsH (_ ::: qry) = case qry of
+listActivatedAccountsH (_ ::: qry) = do
+    json <$> lift (listActivatedAccounts qry)
+
+listActivatedAccounts :: Either (List UserId) (List Handle) -> AppIO [UserAccount]
+listActivatedAccounts = \case
     Left  us -> byIds (fromList us)
     Right hs -> do
-        us <- mapM (lift . API.lookupHandle) (fromList hs)
+        us <- mapM (API.lookupHandle) (fromList hs)
         byIds (catMaybes us)
   where
-    byIds uids = json
-               . filter (isJust . userIdentity . accountUser)
-              <$> (lift $ API.lookupAccounts uids)
+    byIds uids = filter (isJust . userIdentity . accountUser)
+             <$> API.lookupAccounts uids
 
 listAccountsByIdentityH :: JSON ::: Either Email Phone -> Handler Response
 listAccountsByIdentityH (_ ::: emailOrPhone) = lift $ json
     <$> API.lookupAccountsByIdentity emailOrPhone
 
 getActivationCodeH :: JSON ::: Either Email Phone -> Handler Response
 getActivationCodeH (_ ::: emailOrPhone) = do
+    json <$> getActivationCode emailOrPhone
+
+getActivationCode :: Either Email Phone -> Handler GetActivationCodeResp
+getActivationCode emailOrPhone = do
     apair <- lift $ API.lookupActivationCode emailOrPhone
-    maybe (throwStd activationKeyNotFound) (return . found) apair
-  where
-    found (k, c) = json $ object [ "key" .= k, "code" .= c ]
+    maybe (throwStd activationKeyNotFound) (return . GetActivationCodeResp) apair
+
+data GetActivationCodeResp = GetActivationCodeResp (ActivationKey, ActivationCode)
+
+instance ToJSON GetActivationCodeResp where
+    toJSON (GetActivationCodeResp (k, c)) = object [ "key" .= k, "code" .= c ]
 
 getPasswordResetCodeH :: JSON ::: Either Email Phone -> Handler Response
 getPasswordResetCodeH (_ ::: emailOrPhone) = do
+    json <$> getPasswordResetCode emailOrPhone
+
+getPasswordResetCode :: Either Email Phone -> Handler GetPasswordResetCodeResp
+getPasswordResetCode emailOrPhone = do
     apair <- lift $ API.lookupPasswordResetCode emailOrPhone
-    maybe (throwStd invalidPwResetKey) (return . found) apair
-  where
-    found (k, c) = json $ object [ "key" .= k, "code" .= c ]
+    maybe (throwStd invalidPwResetKey) (return . GetPasswordResetCodeResp) apair
```

Comment:
We could return a `Maybe GetPasswordResetCodeResp` here, then we could run this in `AppIO`.

Reply:
efa3ffaf